cloudify_rest_client/resources.py

```diff
@@ -119,57 +121,88 @@ def _upload_deployment_workdir_files(self, uri: str, src_dir: str):
 
         self._save_local_index(src_dir, local_files)
 
-        for file_path, _ in \
+        for file_path, file_mtime in \
                 self._metadata_diff(local_files, manager_files):
             return self._upload_single_file(
                 f'{uri}{file_path}',
                 os.path.join(src_dir, file_path),
+                file_mtime=file_mtime,
             )
 
-    def _upload_single_file(self, uri: str, file_path: str, extract=False):
-        with open(file_path, 'rb') as buffer:
+    def _upload_single_file(self,
+                            uri: str,
+                            file_path: str,
+                            file_rel_path=None,
+                            file_mtime=None,
+                            extract=False):
+        if file_mtime:
+            data = MultipartEncoder(
+                fields={
+                    'file': (
+                        file_rel_path or file_path,
+                        open(file_path, 'rb'),
+                        None,
+                    ),
+                    'mtime': file_mtime,
+                }
+            )
+            request_kwargs = {
+                'data': data,
+                'headers': {'Content-Type': data.content_type},
+            }
+        else:
             # Below is a not very aesthetically pleasing workaround for a known
             # bug in requests, which should be solved in requests 3.x
             # https://github.com/psf/requests/issues/4215
-            data = buffer
-            if _file_is_empty(file_path):
-                data = b''
-
-            try:
-                self.api.put(
-                    uri,
-                    params={'extract': 'yes'} if extract else {},
-                    url_prefix=False,
-                    data=data,
-                )
-            except requests.RequestException as exception:
-                raise CloudifyClientError(
-                    f"Unable to upload deployment's file {file_path} "
-                    f"into {uri}") from exception
+            with open(file_path, 'rb') as buffer:
+                data = buffer
+                if _file_is_empty(file_path):
+                    data = b''
+            request_kwargs = {'data': data}
```

Comment:
but `data` will be already closed by here?

Reply:
yeah, it will/is.  hence push --force